end_to_end/tpu/llama4/Run_Llama4.md

```diff
@@ -19,22 +19,46 @@
 ​Meta's Llama 4 is the latest generation of its open-source large language models (LLMs), unveiled in April 2025. These models are designed to be natively multimodal and multilingual, incorporating a mixture-of-experts (MoE) architecture to enhance performance and efficiency.  The currently supported models are:
 * LLama4 Scout (17B-16E)
 * Llama4 Scout (17B-16E-Instruct)
+* LLama4 Maverick (17B-128E)
+* Llama4 Maverick (17B-128E-Instruct)
 
 
 ## Checkpoint conversion
-Currently, we support converting PyTorch (`.pth`) checkpoints, which can be downloaded from Meta [here](https://www.llama.com/).
+Currently, we support converting both [PyTorch](https://www.llama.com/) and [HuggingFace](https://huggingface.co/collections/meta-llama/llama-4-67f0c30d9fe03840bc9d0164) checkpoints.  Note that we recommend using the `huggingface-cli download` command with environment variable
+`HF_HUB_ENABLE_HF_TRANSFER=1` to download the models.  You can also find a copy of the PyTorch model weights in the HuggingFace repo by looking
+for the models with the `-Original` suffix ([example](https://huggingface.co/meta-llama/Llama-4-Scout-17B-16E-Instruct-Original)).
 
-Once you have downloaded the Scout models, you can run the following command to generate an unscanned checkpoint (preferred for decoding):
+Once you have downloaded the models, you can run the following command to generate an unscanned checkpoint (preferred for decoding):
 
 ```
-JAX_PLATFORMS=CPU python -m MaxText.llama4_ckpt_unscanned --base-model-path [PATH_TO_CHECKPOINT_DIR] --maxtext-model-path [DESIRED_MAXTEXT_CHECKPOINT_OUTPUT_DIR]  --model-size llama4-17b-16e
+JAX_PLATFORMS=CPU python -m MaxText.llama4_ckpt_unscanned --base-model-path [PATH_TO_CHECKPOINT_DIR] --maxtext-model-path [DESIRED_MAXTEXT_CHECKPOINT_OUTPUT_DIR]  --model-size llama4-17b-16e [--huggingface-checkpoint]
 ```
 
 Or the following command to generate a scanned checkpoint (preferred for training):
 ```
 JAX_PLATFORMS=CPU python -m MaxText.llama_or_mistral_ckpt --base-model-path [PATH_TO_CHECKPOINT_DIR] --maxtext-model-path [DESIRED_MAXTEXT_CHECKPOINT_OUTPUT_DIR]  --model-size llama4-17b-16e
 ```
 
+## Pre-training
+You can train from scratch to generate a new checkpoint. One example command to run pretraining with Llama4 Maverick on a v5p-128.
+
+```sh
+python3 -m MaxText.train MaxText/configs/base.yml \
+    base_output_directory=${BASE_OUTPUT_DIRECTORY} \
+    run_name=matmul_pre_training \
+    per_device_batch_size=4 \
+    enable_checkpointing=false \
+    model_name=llama4-17b-128e \
+    ici_fsdp_parallelism=128 \
```

Comment:
v5p-128 has 64 devices.
could you change it to be: 

 `ici_fsdp_parallelism=-1`